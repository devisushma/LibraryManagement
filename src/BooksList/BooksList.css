.booksList
.table {
    margin-top: 10px;
}
.table tr:nth-child(even) {
    background-color: #99bbff;
}
.table td {
    border: 1px solid #fff;
    padding: 0px 10px 0px 10px;
}

