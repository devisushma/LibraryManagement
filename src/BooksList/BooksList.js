import React, { useState, useMemo, useCallback } from 'react';
import {useLocation, useNavigate } from 'react-router-dom';
import './BooksList.css';

const BooksList = ({books, updatedBooks}) => {
    const [checkedBooksId, setCheckedBooks] = useState([])
    const [searchValue, setSearchValue] = useState('')
    const { state } = useLocation();
    const navigate = useNavigate()
    const showCheckbox = state ? state.showCheckbox : false
    const allotedBooksCount = state ? state.allotedBooksCount : 0
    const selectedId = state ? state.selectedId : 0

    const handleSetCheckedBooks = useCallback((e) => {
        const { checked, id } = e.target
        if (checked) setCheckedBooks([...checkedBooksId, id])
        else setCheckedBooks(checkedBooksId.filter(book => id !== book))
    },[checkedBooksId])

    const handleSearch = useCallback((e) => {
        setSearchValue(e.target.value)
    },[])

    const filteredBooks = useMemo(() => {
        if(searchValue === "")  return books
        return books.filter(({isbn, title, author}) => isbn.includes(searchValue) || title.includes(searchValue) || author.includes(searchValue))    
    },[searchValue, books])

    const renderSearchBar = useMemo(() => {
      return (
        <>
        <label>Search: </label>
        <input type="text" onChange={handleSearch} placeholder="Enter text"/>
        </>
      )
    },[handleSearch])

    const handleSubmit = useCallback(() => {
       console.log(checkedBooksId.length, allotedBooksCount)
      if(checkedBooksId.length + allotedBooksCount >3) alert('limit reached')
      else {
        updatedBooks({checkedBooksId, memberId:selectedId})
      } 
      navigate('/members')
    },[checkedBooksId, allotedBooksCount, selectedId, updatedBooks, navigate])

    const renderSubmit = useMemo(() => {
        return (
         showCheckbox ? <button onClick={handleSubmit}>Issue Selected books</button> : <></>
        )
    },[handleSubmit, showCheckbox])

    const renderBooks = useMemo(() => {
        return (
            <table className='table'>
                <tr>
                    {showCheckbox && <th></th>}
                    <th>ISBN</th>
                    <th>Aurthur</th>
                    <th>Title</th>
                    <th>count</th>
                </tr>
                {filteredBooks.filter(({count}) => count>0).map(({ isbn, title, author, count }) => {
                    return (
                        <tr>
                            {showCheckbox && <td><input type="checkbox" id={isbn} onClick={handleSetCheckedBooks} className='booksList__checkbox'/></td> }
                            <td>{isbn}</td>
                            <td>{title}</td>
                            <td>{author}</td>
                            <td>{count}</td>
                        </tr>
                    )
                })}
            </table>
        )
    },[showCheckbox, handleSetCheckedBooks, filteredBooks])

    return (
        <div className='booksList'>
            {renderSearchBar}
            {renderSubmit}
            {renderBooks}
        </div>
    );
}

export default BooksList;