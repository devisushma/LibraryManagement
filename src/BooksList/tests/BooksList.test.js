// import renderer from 'react-test-renderer';
import { render, screen } from '@testing-library/react';
import { BrowserRouter as Router } from "react-router-dom"
import BooksList from '../BooksList';
import {expect, jest, test} from '@jest/globals'

it('renders correctly', () => {
    expect(true).toBeTruthy()
//   const tree = renderer
//     .create(<BooksList />)
//     .toJSON();
//   expect(tree).toMatchSnapshot();
});

it('snapshot correctly', () => {
  const { asFragment } = render(<Router><BooksList books={[]} updatedBooks={[]}/></Router>)
  expect(asFragment()).toMatchSnapshot()
});