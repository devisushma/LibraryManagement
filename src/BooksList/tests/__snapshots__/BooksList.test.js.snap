// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot correctly 1`] = `
<DocumentFragment>
  <div
    class="booksList"
  >
    <label>
      Search: 
    </label>
    <input
      placeholder="Enter text"
      type="text"
    />
    <table
      class="table"
    >
      <tbody>
        <tr>
          <th>
            ISBN
          </th>
          <th>
            Aurthur
          </th>
          <th>
            Title
          </th>
          <th>
            count
          </th>
        </tr>
      </tbody>
    </table>
  </div>
</DocumentFragment>
`;
